Adjusted makefiles to fix cross-compilation

kdmn2 opened the pull request:
Adjusted multiple makefiles to allow for cross-compilation on x86

kdmn2 commented:
It was **clintor** on discord who actually came up with that solution. Otherwise there was samplerate.h error.

frysee commented:
This certainly works, but looks like a toolchain issue. You shouldnt have to duplicate that all over the place. Could you try compiling from https://github.com/LoveRetro/NextUI/tree/samplerate-cleanup with your toolchain changes?

kdmn2 commented:
Unfortunately it still crashes out, but that samplerate error is fixable by adjusting the Dockerfile in the tg5040 toolchain - [PR here](https://github.com/LoveRetro/tg5040-toolchain/pull/1/commits/07f7e70113eea2a4a1ab1399206d583902e25197)

`/opt/aarch64-linux-gnu/bin/aarch64-linux-gnu-gcc minui.c ../common/scaler.c ../common/utils.c ../common/api.c ../../tg5040/platform/platform.c -o build/tg5040/minui.elf -O3 -fomit-frame-pointer -I. -I../common/ -I../../tg5040/platform/ -DPLATFORM=\"tg5040\" -DUSE_SDL2 -Ofast -std=gnu99 -I/usr/include -ldl -lmsettings -flto /usr/lib/aarch64-linux-gnu/libsamplerate.a -lSDL2 -lSDL2_image -lSDL2_ttf -lpthread -lm -lz
aarch64-linux-gnu-gcc: error: /usr/lib/aarch64-linux-gnu/libsamplerate.a: No such file or directory
make[1]: *** [makefile:36: all] Error 1
make[1]: Leaving directory '/root/workspace/all/minui'
make: *** [makefile:21: all] Error 2
root@e9fb19216fb4:~/workspace# cat tg5040/platform/makefile.env
# tg5040
ARCH = -O3
LIBS = -flto /usr/lib/aarch64-linux-gnu/libsamplerate.a
SDL = SDL2root@e9fb19216fb4:~/workspace#
`

kdmn2 commented:
Apologies, I was not on the right branch. Switched to the one you mentioned and I'm hitting the issue that actually prevented me from compiling until I adjusted all those makefiles.

> mkdir -p build/tg5040
> /opt/aarch64-linux-gnu/bin/aarch64-linux-gnu-gcc minui.c ../common/scaler.c ../common/utils.c ../common/api.c ../../tg5040/platform/platform.c -o build/tg5040/minui.elf -O3 -fomit-frame-pointer -I. -I../common/ -I../../tg5040/platform/ -DPLATFORM=\"tg5040\" -DUSE_SDL2 -Ofast -std=gnu99 -ldl -lmsettings -flto -lSDL2 -lSDL2_image -lSDL2_ttf -lpthread -lm -lz
> ../common/api.c:9:10: fatal error: samplerate.h: No such file or directory
>  #include <samplerate.h>
>           ^~~~~~~~~~~~~~
> compilation terminated.
> make[1]: *** [makefile:35: all] Error 1
> make[1]: Leaving directory '/root/workspace/all/minui'

frysee commented:
Apologies for the delayed response, I had to grab some hardware for this and find some time to take an actual look at the cross compilation code. Theres a couple of things I'd like to discuss, but first of all I'd like to ask you to resubmit the PR from a feature branch (i.e. not main). Additionally, while https://github.com/LoveRetro/tg5040-toolchain/pull/1 is open it would make sense to reference it directly from `makefile.toolchain` to ease the review, but I can certainly work around that.

- Whats a bit unclear to me is the mechanics of how your toolchain changes would allow to link proper arm64 dependencies for all the things you are not explicitly pulling down from apt and arent part of the toolchain SDK we are grabbing in `setup-toolchain.sh`. Where do libbluetooth:arm64 and libsqlite3:arm64 come from? I'm by no means a Makefile expert, maybe you could educate me there. I ran a `PLATFORMS=tg5040 make setup tg5040` which seems to succeed, but I'm unsure how much of that is working properly and how much is "Linux has enough fallbacks that this auto-magically-kinda works".
- I assume the proper, idiomatic approach would be to adapt the Dockerfile to export  `PKG_CONFIG_PATH=/usr/lib/aarch64-linux-gnu/pkgconfig` and then use the toolchains `pkg-config --libs` and `--cflags` to resolve the dependencies for us. 
- When building all cores, it seems to error our when compiling snes9x:
```bash
(test ! -f patches/snes9x.patch) || (test -f src/snes9x/.patched) || (cd src/snes9x && git apply -p1 < ../../patches/snes9x.patch && touch .patched && true)
<stdin>:19: trailing whitespace.
   
warning: 1 line adds whitespace errors.
(test ! -d ../../all/cores/patches/snes9x) || (test -f src/snes9x/.patched-all) || (cd src/snes9x &&  touch .patched-all && true)
mkdir -p output
cd src/snes9x/libretro && make  platform=tg5040  -j4
make[2]: Entering directory '/root/workspace/tg5040/cores/src/snes9x/libretro'
/usr/bin/aarch64-linux-gnu-gcc -I../libretro -I../libretro/libretro-common/include -I.. -I../apu/ -I../apu/bapu   -DGIT_VERSION=\"" 49f4845"\" -flto -O3 -Ofast -march=armv8-a+simd -mtune=cortex-a53 -ffast-math -O3 -DNDEBUG -fomit-frame-pointer -Wall -W -Wno-unused-parameter -Wno-missing-field-initializers -fPIC -DRIGHTSHIFT_IS_SAR -D__LIBRETRO__ -DALLOW_CPU_OVERCLOCK -DHAVE_STDINT_H -DHAVE_STRINGS_H -fno-rtti -pedantic  -fno-exceptions -c -o../apu/apu.o ../apu/apu.cpp
/usr/bin/aarch64-linux-gnu-gcc -I../libretro -I../libretro/libretro-common/include -I.. -I../apu/ -I../apu/bapu   -DGIT_VERSION=\"" 49f4845"\" -flto -O3 -Ofast -march=armv8-a+simd -mtune=cortex-a53 -ffast-math -O3 -DNDEBUG -fomit-frame-pointer -Wall -W -Wno-unused-parameter -Wno-missing-field-initializers -fPIC -DRIGHTSHIFT_IS_SAR -D__LIBRETRO__ -DALLOW_CPU_OVERCLOCK -DHAVE_STDINT_H -DHAVE_STRINGS_H -fno-rtti -pedantic  -fno-exceptions -c -o../apu/bapu/dsp/sdsp.o ../apu/bapu/dsp/sdsp.cpp
/usr/bin/aarch64-linux-gnu-gcc -I../libretro -I../libretro/libretro-common/include -I.. -I../apu/ -I../apu/bapu   -DGIT_VERSION=\"" 49f4845"\" -flto -O3 -Ofast -march=armv8-a+simd -mtune=cortex-a53 -ffast-math -O3 -DNDEBUG -fomit-frame-pointer -Wall -W -Wno-unused-parameter -Wno-missing-field-initializers -fPIC -DRIGHTSHIFT_IS_SAR -D__LIBRETRO__ -DALLOW_CPU_OVERCLOCK -DHAVE_STDINT_H -DHAVE_STRINGS_H -fno-rtti -pedantic  -fno-exceptions -c -o../apu/bapu/smp/smp.o ../apu/bapu/smp/smp.cpp
/bin/sh: 1: /usr/bin/aarch64-linux-gnu-gcc: not found
/bin/sh: 1: /usr/bin/aarch64-linux-gnu-gcc: not found
/usr/bin/aarch64-linux-gnu-gcc -I../libretro -I../libretro/libretro-common/include -I.. -I../apu/ -I../apu/bapu   -DGIT_VERSION=\"" 49f4845"\" -flto -O3 -Ofast -march=armv8-a+simd -mtune=cortex-a53 -ffast-math -O3 -DNDEBUG -fomit-frame-pointer -Wall -W -Wno-unused-parameter -Wno-missing-field-initializers -fPIC -DRIGHTSHIFT_IS_SAR -D__LIBRETRO__ -DALLOW_CPU_OVERCLOCK -DHAVE_STDINT_H -DHAVE_STRINGS_H -fno-rtti -pedantic  -fno-exceptions -c -o../apu/bapu/smp/smp_state.o ../apu/bapu/smp/smp_state.cpp
/bin/sh: 1: /usr/bin/aarch64-linux-gnu-gcc: not found
make[2]: *** [Makefile:731: ../apu/apu.o] Error 127
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [Makefile:731: ../apu/bapu/dsp/sdsp.o] Error 127
make[2]: *** [Makefile:731: ../apu/bapu/smp/smp.o] Error 127
/bin/sh: 1: /usr/bin/aarch64-linux-gnu-gcc: not found
make[2]: *** [Makefile:731: ../apu/bapu/smp/smp_state.o] Error 127
make[2]: Leaving directory '/root/workspace/tg5040/cores/src/snes9x/libretro'
make[1]: *** [../../all/cores/makefile:72: output/snes9x_libretro.so] Error 2
make[1]: Leaving directory '/root/workspace/tg5040/cores'
make: *** [makefile:35: all] Error 2
make[2]: *** [makefile.toolchain:29: build] Error 2
make[1]: *** [makefile:42: build] Error 2
make: *** [makefile:189: tg5040] Error 1
```

clintonium-119 commented:
I'll have to give this a try.  Also hitting the `samplerate.h` error

clintonium-119 commented:
So I've tried this out for an hour or two.  I'm getting past the samplerate.h errors, but am having a different errors crop up later in the build process.  Haven't had time to debug very deeply, and probably won't for a while.

rudism commented:
I took a (failed so far) stab at getting cross compile to work too. I have a few notes:

- All of the include flags can be set one time in [`workspace/tg5040/platform/makefile.env`](https://github.com/LoveRetro/NextUI/blob/main/workspace/tg5040/platform/makefile.env) instead of having to do it in each makefile directly.
- Over in [tg5040-toolchain](https://github.com/LoveRetro/tg5040-toolchain) I needed several additional changes:
  - Building `zlib` in the `Dockerfile` is done on the host platform and so isn't found as a dependency for cross builds, I needed to install `gcc-aarch64-linux-gnu`, `binutils-aarch64-linux-gnu`, and make a `platform.cmake` file instructing it to cross build for aarch64 (which still failed until I installed a couple more dependencies for arm64 too, I forget which ones now)... It also needed some PKG_CONFIG details pointing it to `aarch64` locations to get past some errors.
  - There are some makefiles that are copying hard-coded paths to various libs around which don't exist at the expected path when cross compiled (needed to change `/usr/lib/` to `/usr/aarch64-linux-gnu/lib`, for example).
- I needed to add some `LDFLAGS` to the `makefile.env` I mentioned above because the linker wasn't looking in all the necessary `aarch64` lib directories

All of the above got me to a (messy) point where the toolchain was at least successfully cross building, but building the system stuff and cores hit some errors and I didn't have time to keep investigating or tweaking. I'm pretty sure the changes I made almost definitely would have broken the build on native arm devices... I'm not knowledgable enough in makefiles and c/c++ builds to know how to clean it up and get it working universally, but I learned enough that I don't think it's going to be trivial to get cross builds fully working.

ro8inmorgan commented:
I fixed crosscompile you can check the branch

working_x86crosscompile 

Make sure to also use this branch on our toolchain. 

It will fully compile, but still need to check if it doesnt break arm builds before I merge. But for now you can use this branch on both repo's (NextUI and our Toolchain repo) to compile on x86

frysee commented:
will be fixed with #470
